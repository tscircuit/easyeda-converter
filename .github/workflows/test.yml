name: Test and Benchmark

on:
  pull_request:
    branches:
      - main

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout code
        uses: actions/checkout@v2

      - name: Use bun
        uses: oven-sh/setup-bun@v1

      - name: Install dependencies
        run: bun i

      - name: Test
        run: bun test --timeout 60_000
