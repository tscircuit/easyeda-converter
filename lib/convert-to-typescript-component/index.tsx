import type { AnyCircuitElement } from "circuit-json"
import {
  EasyEdaJsonSchema,
  type BetterEasyEdaJson,
} from "lib/schemas/easy-eda-json-schema"
import { su } from "@tscircuit/soup-util"
import { soupTypescriptComponentTemplate } from "./soup-typescript-component-template"
import {
  convertEasyEdaJsonToCircuitJson,
  convertEasyEdaJsonToTscircuitSoupJson,
} from "lib/convert-easyeda-json-to-tscircuit-soup-json"
import { normalizeManufacturerPartNumber } from "lib/utils/normalize-manufacturer-part-number"
import { normalizePinLabels } from "lib/normalize-pin-labels"

export const convertRawEasyToTsx = async (rawEasy: any) => {
  const betterEasy = EasyEdaJsonSchema.parse(rawEasy)
  const result = await convertBetterEasyToTsx({
    betterEasy,
  })
  return result
}

export const convertBetterEasyToTsx = async ({
  betterEasy,
}: {
  betterEasy: BetterEasyEdaJson
}): Promise<string> => {
  const circuitJson = convertEasyEdaJsonToCircuitJson(betterEasy, {
    useModelCdn: true,
    shouldRecenter: true,
  })
  const rawPn = betterEasy.dataStr.head.c_para["Manufacturer Part"]
  const pn = normalizeManufacturerPartNumber(rawPn)
  const [cad_component] = su(circuitJson).cad_component.list()

  const sourcePorts = su(circuitJson).source_port.list()

  const pinLabels: Record<string, string[]> = {}
  for (const sourcePort of sourcePorts) {
    pinLabels[sourcePort.name] = [
      sourcePort.name,
      ...(sourcePort.port_hints ?? []),
    ]
  }

  let modelObjUrl: string | undefined
  if (cad_component.model_obj_url) {
    const isValidUrl = await checkModelObjUrlValidity(
      cad_component.model_obj_url,
    )
    if (isValidUrl) {
      modelObjUrl = cad_component.model_obj_url
    }
  }

  const supplierPartNumbers: Record<string, string[]> = {
    lcsc: [betterEasy.lcsc.number],
  }

  return soupTypescriptComponentTemplate({
    componentName: pn,
    pinLabels,

    objUrl: modelObjUrl,
    circuitJson,
    supplierPartNumbers,
  })
}

const checkModelObjUrlValidity = async (url: string): Promise<boolean> => {
  try {
    const response = await fetch(url, { method: "HEAD" })
    return response.status === 200
  } catch (error) {
    console.error(`Error checking model object URL ${url}:`, error)
    return false
  }
}
