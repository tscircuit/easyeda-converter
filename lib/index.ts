export * from "./convert-easyeda-json-to-tscircuit-soup-json"
export * from "./fetch-easyeda-json"
export { convertRawEasyEdaToTs } from "./convert-to-typescript-component"
export { normalizeManufacturerPartNumber } from "./utils/normalize-manufacturer-part-number"
export * from "./schemas/easy-eda-json-schema"
export { convertEasyEdaJsonToVariousFormats } from "./convert-easyeda-json-to-various-formats"
