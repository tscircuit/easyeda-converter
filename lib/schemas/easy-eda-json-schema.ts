import { z } from "zod"
import {
  PackageDetailShapeSchema,
  ShapeItemSchema,
} from "./package-detail-shape-schema"
import { SingleLetterShapeSchema } from "./single-letter-shape-schema"

export const maybeNumber = z
  .any()
  .transform((k) => (k === "nan" || Number.isNaN(k) ? null : k))
  .pipe(z.number().nullable().optional())

export const SzlcscSchema = z.object({
  id: z.number(),
  number: z.string(),
  step: z.number().optional(),
  min: z.number().optional(),
  price: z.number().optional(),
  stock: z.number().optional(),
  url: z.string().url().optional(),
  image: z.string().optional().optional(),
})

export const LcscSchema = z.object({
  id: z.number(),
  number: z.string(),
  step: z.number().optional(),
  min: z.number().optional(),
  price: z.number().optional(),
  stock: z.number().optional(),
  url: z.string().url().optional(),
})

export const OwnerSchema = z.object({
  uuid: z.string(),
  username: z.string(),
  nickname: z.string(),
  avatar: z.string(),
})

export const HeadSchema = z.object({
  docType: z.preprocess((val) => (val == null ? val : String(val)), z.string()),
  editorVersion: z.string(),
  c_para: z.record(z.string(), z.string().nullable()),
  x: z.number(),
  y: z.number(),
  puuid: z.string().optional(),
  uuid: z.string(),
  utime: z.preprocess((val) => {
    if (val === "") return 0
    if (typeof val === "string") return Number(val)
    return val
  }, z.number()),
  importFlag: z.number().optional(),
  c_spiceCmd: z.any().optional(),
  hasIdFlag: z.boolean(),
})

export const BBoxSchema = z.object({
  x: z.number(),
  y: z.number(),
  width: z.number(),
  height: z.number(),
})

export const LayerItemSchema = z.object({
  name: z.string(),
  color: z.string(),
  visible: z.boolean(),
  active: z.boolean(),
  config: z.boolean(),
  transparency: z.boolean(),
})

export const ObjectItemSchema = z.object({
  name: z.string(),
  visible: z.boolean(),
  locked: z.boolean(),
})

export const DataStrSchema = z.object({
  head: HeadSchema,
  canvas: z.string(),
  shape: z.array(SingleLetterShapeSchema),
  BBox: BBoxSchema,
  colors: z.array(z.unknown()),
})

export const PackageDetailDataStrSchema = z.object({
  head: HeadSchema,
  canvas: z.string(),
  shape: z
    .array(z.string())
    .transform((shapes) =>
      shapes.map((shape) => {
        const [type, ...data] = shape.split("~")
        return ShapeItemSchema.parse({ type, data: data.join("~") })
      }),
    )
    .pipe(z.array(PackageDetailShapeSchema)),
  layers: z.array(z.string()).transform((layers) =>
    layers.map((layer) => {
      const [name, color, visible, active, config, transparency] =
        layer.split("~")
      return LayerItemSchema.parse({
        name,
        color,
        visible: visible === "true",
        active: active === "true",
        config: config === "true",
        transparency: transparency === "true",
      })
    }),
  ),
  objects: z.array(z.string()).transform((objects) =>
    objects.map((obj) => {
      const [name, visible, locked] = obj.split("~")
      return ObjectItemSchema.parse({
        name,
        visible: visible === "true",
        locked: locked === "true",
      })
    }),
  ),
  BBox: BBoxSchema,
  netColors: z.array(z.unknown()).optional(),
})

export const PackageDetailSchema = z.object({
  uuid: z.string(),
  title: z.string(),
  docType: z.number(),
  updateTime: z.number(),
  owner: OwnerSchema,
  datastrid: z.string(),
  writable: z.boolean(),
  dataStr: PackageDetailDataStrSchema,
})

export const EasyEdaJsonSchema = z.object({
  uuid: z.string(),
  title: z.string(),
  description: z.string(),
  docType: z.number(),
  type: z.number(),
  szlcsc: SzlcscSchema,
  lcsc: LcscSchema,
  owner: OwnerSchema,
  tags: z.array(z.string()),
  updateTime: z.number(),
  updated_at: z.string(),
  dataStr: DataStrSchema,
  verify: z.boolean(),
  SMT: z.boolean().optional(),
  datastrid: z.string(),
  jlcOnSale: z.number().optional(),
  writable: z.boolean(),
  isFavorite: z.boolean(),
  packageDetail: PackageDetailSchema,
})

export type RawEasyEdaJson = z.input<typeof EasyEdaJsonSchema>
export type BetterEasyEdaJson = z.infer<typeof EasyEdaJsonSchema>
