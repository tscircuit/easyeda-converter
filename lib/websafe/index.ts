export * from "./convert-to-typescript-component"
export * from "./fetch-easyeda-json"
export * from "./generate-footprint-tsx"
export * from "./get-model-cdn-url"
