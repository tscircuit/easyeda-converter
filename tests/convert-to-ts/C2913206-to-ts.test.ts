import { it, expect } from "bun:test"
import chipRawEasy from "../assets/C2913206.raweasy.json"
import { convertBetterEasyToTsx } from "lib/websafe/convert-to-typescript-component"
import { EasyEdaJsonSchema } from "lib/schemas/easy-eda-json-schema"
import { convertEasyEdaJsonToCircuitJson } from "lib/convert-easyeda-json-to-tscircuit-soup-json"
import { su } from "@tscircuit/circuit-json-util"
import type { PcbSmtPadRect } from "circuit-json"

it("should convert C2913206 into typescript file", async () => {
  const betterEasy = EasyEdaJsonSchema.parse(chipRawEasy)

  const circuitJson = convertEasyEdaJsonToCircuitJson(betterEasy)

  const smtPad1 = su(circuitJson)
    .pcb_smtpad.list()
    .find((smtpad) => smtpad.port_hints?.includes("pin1"))! as PcbSmtPadRect

  expect(smtPad1.width > smtPad1.height).toBeTrue()

  const result = await convertBetterEasyToTsx({
    betterEasy,
  })

  expect(result).not.toContain("milmm")
  expect(result).not.toContain("NaNmm")

  // Add more specific assertions here based on the component

  expect(result).toMatchInlineSnapshot(`
    "import type { ChipProps } from "@tscircuit/props"

    const pinLabels = {
      pin1: ["GND1"],
      pin2: ["GND2"],
      pin3: ["3V3"],
      pin4: ["IO0"],
      pin5: ["IO1"],
      pin6: ["IO2"],
      pin7: ["IO3"],
      pin8: ["IO4"],
      pin9: ["IO5"],
      pin10: ["IO6"],
      pin11: ["IO7"],
      pin12: ["IO8"],
      pin13: ["IO9"],
      pin14: ["IO10"],
      pin15: ["IO11"],
      pin16: ["IO12"],
      pin17: ["IO13"],
      pin18: ["IO14"],
      pin19: ["IO15"],
      pin20: ["IO16"],
      pin21: ["IO17"],
      pin22: ["IO18"],
      pin23: ["IO19"],
      pin24: ["IO20"],
      pin25: ["IO21"],
      pin26: ["IO26"],
      pin27: ["IO47"],
      pin28: ["IO33"],
      pin29: ["IO34"],
      pin30: ["IO48"],
      pin31: ["IO35"],
      pin32: ["IO36"],
      pin33: ["IO37"],
      pin34: ["IO38"],
      pin35: ["IO39"],
      pin36: ["IO40"],
      pin37: ["IO41"],
      pin38: ["IO42"],
      pin39: ["TXD0"],
      pin40: ["RXD0"],
      pin41: ["IO45"],
      pin42: ["GND3"],
      pin43: ["GND4"],
      pin44: ["IO46"],
      pin45: ["EN"],
      pin46: ["GND5"],
      pin47: ["GND6"],
      pin48: ["GND7"],
      pin49: ["GND8"],
      pin50: ["GND9"],
      pin51: ["GND10"],
      pin52: ["GND11"],
      pin53: ["GND12"],
      pin54: ["GND13"],
      pin55: ["GND14"],
      pin56: ["GND15"],
      pin57: ["GND16"],
      pin58: ["GND17"],
      pin59: ["GND18"],
      pin60: ["GND19"],
      pin61: ["GND20"],
      pin62: ["GND21"],
      pin63: ["GND22"],
      pin64: ["GND23"],
      pin65: ["GND24"],
      pin66: ["GND25"],
      pin67: ["GND26"],
      pin68: ["GND27"],
      pin69: ["GND28"],
      pin70: ["GND29"],
      pin71: ["GND30"],
      pin72: ["GND31"],
      pin73: ["GND32"]
    } as const

    export const ESP32_S3_MINI_1_N8 = (props: ChipProps<typeof pinLabels>) => {
      return (
        <chip
          pinLabels={pinLabels}
          supplierPartNumbers={{
      "jlcpcb": [
        "C2913206"
      ]
    }}
          manufacturerPartNumber="ESP32_S3_MINI_1_N8"
          footprint={<footprint>
            <smtpad portHints={["pin1"]} pcbX="-6.975094000000013mm" pcbY="5.950076999999965mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin2"]} pcbX="-6.975094000000013mm" pcbY="5.10019299999999mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin3"]} pcbX="-6.975094000000013mm" pcbY="4.250054999999975mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin4"]} pcbX="-6.975094000000013mm" pcbY="3.4001710000000003mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin5"]} pcbX="-6.975094000000013mm" pcbY="2.550032999999985mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin6"]} pcbX="-6.975094000000013mm" pcbY="1.7001489999998967mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin7"]} pcbX="-6.975094000000013mm" pcbY="0.8500109999998813mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin8"]} pcbX="-6.975094000000013mm" pcbY="0.0001269999999067295mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin9"]} pcbX="-6.975094000000013mm" pcbY="-0.8497570000000678mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin10"]} pcbX="-6.975094000000013mm" pcbY="-1.6998950000000832mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin11"]} pcbX="-6.975094000000013mm" pcbY="-2.5497790000001714mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin12"]} pcbX="-6.975094000000013mm" pcbY="-3.399917000000073mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin13"]} pcbX="-6.975094000000013mm" pcbY="-4.249801000000048mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin14"]} pcbX="-6.975094000000013mm" pcbY="-5.099939000000063mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin15"]} pcbX="-6.975094000000013mm" pcbY="-5.949823000000038mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin16"]} pcbX="-5.949950000000058mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin17"]} pcbX="-5.1000660000000835mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin18"]} pcbX="-4.249928000000068mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin19"]} pcbX="-3.4000440000000935mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin20"]} pcbX="-2.549906000000078mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin21"]} pcbX="-1.7000220000001036mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin22"]} pcbX="-0.8498839999999745mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin23"]} pcbX="0mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin24"]} pcbX="0.8498839999999745mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin25"]} pcbX="1.70002199999999mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin26"]} pcbX="2.5499059999999645mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin27"]} pcbX="3.40004399999998mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin28"]} pcbX="4.249927999999954mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin29"]} pcbX="5.10006599999997mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin30"]} pcbX="5.949949999999944mm" pcbY="-7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin31"]} pcbX="6.975094000000013mm" pcbY="-5.949823000000038mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin32"]} pcbX="6.975094000000013mm" pcbY="-5.099939000000063mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin33"]} pcbX="6.975094000000013mm" pcbY="-4.249801000000048mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin34"]} pcbX="6.975094000000013mm" pcbY="-3.399917000000073mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin35"]} pcbX="6.975094000000013mm" pcbY="-2.5497790000001714mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin36"]} pcbX="6.975094000000013mm" pcbY="-1.6998950000000832mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin37"]} pcbX="6.975094000000013mm" pcbY="-0.8497570000000678mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin38"]} pcbX="6.975094000000013mm" pcbY="0.0001269999999067295mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin39"]} pcbX="6.975094000000013mm" pcbY="0.8500109999998813mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin40"]} pcbX="6.975094000000013mm" pcbY="1.7001489999998967mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin41"]} pcbX="6.975094000000013mm" pcbY="2.550032999999985mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin42"]} pcbX="6.975094000000013mm" pcbY="3.4001710000000003mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin43"]} pcbX="6.975094000000013mm" pcbY="4.250054999999975mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin44"]} pcbX="6.975094000000013mm" pcbY="5.10019299999999mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin45"]} pcbX="6.975094000000013mm" pcbY="5.950076999999965mm" width="0.7999983999999999mm" height="0.39999919999999994mm" shape="rect" />
    <smtpad portHints={["pin46"]} pcbX="5.949949999999944mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin47"]} pcbX="5.10006599999997mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin48"]} pcbX="4.249927999999954mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin49"]} pcbX="3.40004399999998mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin50"]} pcbX="2.5499059999999645mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin51"]} pcbX="1.70002199999999mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin52"]} pcbX="0.8498839999999745mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin53"]} pcbX="0mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin54"]} pcbX="-0.8498839999999745mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin55"]} pcbX="-1.7000220000001036mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin56"]} pcbX="-2.549906000000078mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin57"]} pcbX="-3.4000440000000935mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin58"]} pcbX="-4.249928000000068mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin59"]} pcbX="-5.1000660000000835mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin60"]} pcbX="-5.949950000000058mm" pcbY="7.025004999999965mm" width="0.39999919999999994mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin61"]} pcbX="-6.999986000000149mm" pcbY="7.000113000000056mm" width="0.7999983999999999mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin62"]} pcbX="6.999986000000035mm" pcbY="7.000113000000056mm" width="0.7999983999999999mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin63"]} pcbX="-6.999986000000149mm" pcbY="-6.999859000000129mm" width="0.7999983999999999mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin64"]} pcbX="6.999986000000035mm" pcbY="-6.999859000000129mm" width="0.7999983999999999mm" height="0.7999983999999999mm" shape="rect" />
    <smtpad portHints={["pin65"]} pcbX="-1.6499840000000177mm" pcbY="1.6501109999999244mm" width="1.1999975999999999mm" height="1.1999975999999999mm" shape="rect" />
    <smtpad portHints={["pin66"]} pcbX="0mm" pcbY="1.6501109999999244mm" width="1.1999975999999999mm" height="1.1999975999999999mm" shape="rect" />
    <smtpad portHints={["pin67"]} pcbX="1.649983999999904mm" pcbY="1.6501109999999244mm" width="1.1999975999999999mm" height="1.1999975999999999mm" shape="rect" />
    <smtpad portHints={["pin68"]} pcbX="-1.6499840000000177mm" pcbY="0.0001269999999067295mm" width="1.1999975999999999mm" height="1.1999975999999999mm" shape="rect" />
    <smtpad portHints={["pin69"]} pcbX="0mm" pcbY="0.0001269999999067295mm" width="1.1999975999999999mm" height="1.1999975999999999mm" shape="rect" />
    <smtpad portHints={["pin70"]} pcbX="1.649983999999904mm" pcbY="0.0001269999999067295mm" width="1.1999975999999999mm" height="1.1999975999999999mm" shape="rect" />
    <smtpad portHints={["pin71"]} pcbX="-1.6499840000000177mm" pcbY="-1.6498569999999972mm" width="1.1999975999999999mm" height="1.1999975999999999mm" shape="rect" />
    <smtpad portHints={["pin72"]} pcbX="0mm" pcbY="-1.6498569999999972mm" width="1.1999975999999999mm" height="1.1999975999999999mm" shape="rect" />
    <smtpad portHints={["pin73"]} pcbX="1.649983999999904mm" pcbY="-1.6498569999999972mm" width="1.1999975999999999mm" height="1.1999975999999999mm" shape="rect" />
    <silkscreenpath route={[{"x":-4.8999902000000475,"y":12.100128199999858},{"x":-3.3999932000001536,"y":12.100128199999858},{"x":-3.3999932000001536,"y":10.500131399999987},{"x":-1.1999976000000743,"y":10.500131399999987},{"x":-1.1999976000000743,"y":12.100128199999858},{"x":0.9999979999998914,"y":12.100128199999858},{"x":0.9999979999998914,"y":10.500131399999987},{"x":3.1999935999999707,"y":10.500131399999987},{"x":3.1999935999999707,"y":12.100128199999858},{"x":5.300014799999872,"y":12.100128199999858},{"x":5.300014799999872,"y":9.400133600000004}]} />
    <silkscreenpath route={[{"x":-6.50001240000006,"y":9.400133600000004},{"x":-6.50001240000006,"y":12.100128199999858},{"x":-4.8999902000000475,"y":12.100128199999858},{"x":-4.8999902000000475,"y":9.400133600000004}]} />
    <silkscreenpath route={[{"x":-7.70001000000002,"y":7.749743199999898},{"x":7.800009799999998,"y":7.749743199999898},{"x":7.800009799999998,"y":7.786014399999885}]} />
    <silkscreenpath route={[{"x":-7.70001000000002,"y":12.800126800000044},{"x":7.800009799999998,"y":12.800126800000044},{"x":7.800009799999998,"y":-7.6998830000001135},{"x":-7.70001000000002,"y":-7.6998830000001135},{"x":-7.70001000000002,"y":12.800126800000044}]} />
          </footprint>}
          cadModel={{
            objUrl: "https://modelcdn.tscircuit.com/easyeda_models/download?uuid=1270b5cf7aa247fc9d0ae79a19686940&pn=C2913206",
            rotationOffset: { x: 0, y: 0, z: 0 },
            positionOffset: { x: 0, y: 0, z: 0 },
          }}
          {...props}
        />
      )
    }"
  `)
})
