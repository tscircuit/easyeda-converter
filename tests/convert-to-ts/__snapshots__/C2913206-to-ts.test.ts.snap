// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should convert C2913206 into typescript file 1`] = `
"const pinLabels = {
  "pin1": [
    "pin1",
    "GND1"
  ],
  "pin2": [
    "pin2",
    "GND2"
  ],
  "pin3": [
    "pin3",
    "3V3"
  ],
  "pin4": [
    "pin4",
    "IO0"
  ],
  "pin5": [
    "pin5",
    "IO1"
  ],
  "pin6": [
    "pin6",
    "IO2"
  ],
  "pin7": [
    "pin7",
    "IO3"
  ],
  "pin8": [
    "pin8",
    "IO4"
  ],
  "pin9": [
    "pin9",
    "IO5"
  ],
  "pin10": [
    "pin10",
    "IO6"
  ],
  "pin11": [
    "pin11",
    "IO7"
  ],
  "pin12": [
    "pin12",
    "IO8"
  ],
  "pin13": [
    "pin13",
    "IO9"
  ],
  "pin14": [
    "pin14",
    "IO10"
  ],
  "pin15": [
    "pin15",
    "IO11"
  ],
  "pin16": [
    "pin16",
    "IO12"
  ],
  "pin17": [
    "pin17",
    "IO13"
  ],
  "pin18": [
    "pin18",
    "IO14"
  ],
  "pin19": [
    "pin19",
    "IO15"
  ],
  "pin20": [
    "pin20",
    "IO16"
  ],
  "pin21": [
    "pin21",
    "IO17"
  ],
  "pin22": [
    "pin22",
    "IO18"
  ],
  "pin23": [
    "pin23",
    "IO19"
  ],
  "pin24": [
    "pin24",
    "IO20"
  ],
  "pin25": [
    "pin25",
    "IO21"
  ],
  "pin26": [
    "pin26",
    "IO26"
  ],
  "pin27": [
    "pin27",
    "IO47"
  ],
  "pin28": [
    "pin28",
    "IO33"
  ],
  "pin29": [
    "pin29",
    "IO34"
  ],
  "pin30": [
    "pin30",
    "IO48"
  ],
  "pin31": [
    "pin31",
    "IO35"
  ],
  "pin32": [
    "pin32",
    "IO36"
  ],
  "pin33": [
    "pin33",
    "IO37"
  ],
  "pin34": [
    "pin34",
    "IO38"
  ],
  "pin35": [
    "pin35",
    "IO39"
  ],
  "pin36": [
    "pin36",
    "IO40"
  ],
  "pin37": [
    "pin37",
    "IO41"
  ],
  "pin38": [
    "pin38",
    "IO42"
  ],
  "pin39": [
    "pin39",
    "TXD0"
  ],
  "pin40": [
    "pin40",
    "RXD0"
  ],
  "pin41": [
    "pin41",
    "IO45"
  ],
  "pin42": [
    "pin42",
    "GND3"
  ],
  "pin43": [
    "pin43",
    "GND4"
  ],
  "pin44": [
    "pin44",
    "IO46"
  ],
  "pin45": [
    "pin45",
    "EN"
  ],
  "pin46": [
    "pin46",
    "GND5"
  ],
  "pin47": [
    "pin47",
    "GND6"
  ],
  "pin48": [
    "pin48",
    "GND7"
  ],
  "pin49": [
    "pin49",
    "GND8"
  ],
  "pin50": [
    "pin50",
    "GND9"
  ],
  "pin51": [
    "pin51",
    "GND10"
  ],
  "pin52": [
    "pin52",
    "GND11"
  ],
  "pin53": [
    "pin53",
    "GND12"
  ],
  "pin54": [
    "pin54",
    "GND13"
  ],
  "pin55": [
    "pin55",
    "GND14"
  ],
  "pin56": [
    "pin56",
    "GND15"
  ],
  "pin57": [
    "pin57",
    "GND16"
  ],
  "pin58": [
    "pin58",
    "GND17"
  ],
  "pin59": [
    "pin59",
    "GND18"
  ],
  "pin60": [
    "pin60",
    "GND19"
  ],
  "pin61": [
    "pin61",
    "GND20"
  ],
  "pin62": [
    "pin62",
    "GND21"
  ],
  "pin63": [
    "pin63",
    "GND22"
  ],
  "pin64": [
    "pin64",
    "GND23"
  ],
  "pin65": [
    "pin65",
    "GND24"
  ],
  "pin66": [
    "pin66",
    "GND25"
  ],
  "pin67": [
    "pin67",
    "GND26"
  ],
  "pin68": [
    "pin68",
    "GND27"
  ],
  "pin69": [
    "pin69",
    "GND28"
  ],
  "pin70": [
    "pin70",
    "GND29"
  ],
  "pin71": [
    "pin71",
    "GND30"
  ],
  "pin72": [
    "pin72",
    "GND31"
  ],
  "pin73": [
    "pin73",
    "GND32"
  ]
}"
`;
